// stock-symbol-lookup/src/ftpUtils.ts

import * as Client from 'ftp';

var ftpUtils: any = {
    fetchFile: function (file: any): Promise<any> {
        var ftpClient = new (<any>Client)(),
            protocol: string = 'ftp://',
            host: string = 'ftp.nasdaqtrader.com',
            ftpDir: string = '/SymbolDirectory/',
            ftpOptions: any = {
                host: host
            };

        console.log('Fetch ' + file);

        return new Promise((resolve, reject) => {
            ftpClient.on('ready', () => {
                let data: string = '';

                ftpClient.get(ftpDir + file, (err: any, stream: any) => {
                    if (err) {
                        reject(err);
                    }

                    stream.once('close', () => {
                        ftpClient.end();
                        resolve(data);
                    });

                    stream.on('data', (buffer: any) => {
                        data += buffer.toString();
                    });
                });
            });
            ftpClient.connect(ftpOptions);
        })
    }
};

export {ftpUtils};
