// stock-symbol-lookup/built/index.js
"use strict";
var _this = this;
var dataUtils_1 = require("./dataUtils");
var searchUtils_1 = require("./searchUtils");
var StockSymbolLookup = {
    data: {},
    loadData: function () {
        return dataUtils_1.dataUtils.getData()
            .then(function (data) {
            return _this.data = searchUtils_1.searchUtils.data = data;
        })
            .catch(function (err) {
            console.log(err);
        });
    },
    getData: function () {
        return _this.data;
    },
    setData: function (data) {
        return _this.data = searchUtils_1.searchUtils.data = data;
    },
    search: function (input, maxEntries, unique) {
        return searchUtils_1.searchUtils.search(input, maxEntries, unique)
            .then(function (results) {
            return {
                symbols: results[0],
                securities: results[1]
            };
        })
            .catch(function (err) {
            console.log(err);
        });
    },
    searchAll: function (input, maxEntries) {
        return searchUtils_1.searchUtils.searchAll(input, maxEntries)
            .then(function (results) {
            return results;
        })
            .catch(function (err) {
            console.log(err);
        });
    },
    searchBySymbol: function (input, maxEntries) {
        return searchUtils_1.searchUtils.searchBySymbol(input, maxEntries)
            .then(function (results) {
            return results;
        })
            .catch(function (err) {
            console.log(err);
        });
    },
    searchBySecurity: function (input, maxEntries) {
        return searchUtils_1.searchUtils.searchBySecurity(input, maxEntries)
            .then(function (results) {
            return results;
        })
            .catch(function (err) {
            console.log(err);
        });
    }
};
module.exports = StockSymbolLookup;
